src/span-data.ts

```diff
@@ -125,16 +145,29 @@ export class RootSpanData extends BaseSpanData implements types.RootSpan {
   createChildSpan(options?: SpanOptions): Span {
     options = options || {name: ''};
     const skipFrames = options.skipFrames ? options.skipFrames + 1 : 1;
-    return new ChildSpanData(
+    const child = new ChildSpanData(
         this.trace,       /* Trace object */
         options.name,     /* Span name */
         this.span.spanId, /* Parent's span ID */
         skipFrames);      /* # of frames to skip in stack trace */
+    this.children.push(child);
+    return child;
   }
 
   endSpan(timestamp?: Date) {
+    if (!!this.span.endTime) {
+      return;
+    }
     super.endSpan(timestamp);
     traceWriter.get().writeTrace(this.trace);
+    this.children.forEach(child => {
+      if (!child.span.endTime) {
+        // Child hasn't ended yet.
+        // Override the endSpan function to make it re-publish only itself.
+        child.endSpan = RootSpanData.overrideChildEndSpanHandler;
```

Comment:
Is there another way to do this instead of monkey patching? Could you have a test in the `endSpan` function? It is best to avoid monkey patching functions unless absolutely necessary. A reader will have to know that the `endSpan` method is part of the dynamic state rather than just the data.

Reply:
I've added a commit that changes the way this is being done, LMK what you think.